File: src/StaticQueue.zig
const std = @import("std");

pub fn StaticQueue(comptime T: type, comptime capacity: usize) type {
    return struct {
        const Self = @This();
        values: [capacity]T = undefined,
        capacity: usize = capacity,
        begin: usize = 0,
        end: usize = 0,

        pub fn push(queue: *Self, value: T) !void {
            if (queue.size() == capacity) {
                return error.OutOfMemory;
            }
            queue.values[queue.end] = value;
            queue.end = (queue.end + 1) % capacity;
        }

        pub fn pop(queue: *Self) ?T {
            if (queue.size() == 0) {
                return null;
            }
            const value = queue.values[queue.begin];
            queue.begin = (queue.begin + 1) % capacity;
            return value;
        }

        pub fn size(queue: *Self) usize {
            return (queue.end + capacity - queue.begin) % capacity;
        }
    };
}

File: ts/src/staticQueue.ts
export class StaticQueue {
  private readonly values: Int32Array;
  private begin = 0;
  private end = 0;
  private length = 0;

  constructor(private readonly capacity: number) {
    if (!Number.isInteger(capacity) || capacity <= 0) {
      throw new RangeError("StaticQueue: capacity must be positive integer");
    }
    this.values = new Int32Array(capacity);
  }

  push(value: number): void {
    if (this.length === this.capacity) {
      throw new RangeError("StaticQueue overflow");
    }
    this.values[this.end] = value;
    this.end = (this.end + 1) % this.capacity;
    this.length += 1;
  }

  pop(): number | null {
    if (this.length === 0) return null;
    const value = this.values[this.begin];
    this.begin = (this.begin + 1) % this.capacity;
    this.length -= 1;
    return value ?? null;
  }

  size(): number {
    return this.length;
  }

  reset(): void {
    this.begin = 0;
    this.end = 0;
    this.length = 0;
  }
}
